src/contexts/UserState.tsx

```diff
@@ -74,38 +81,85 @@ const initialUserContext: {
   setUserState: React.Dispatch<UserActions>
   onLogin: OnLogin
   logout: () => void
+  refreshJWT: () => void
 } = {
   userState: initialState,
   setUserState: () => {},
   onLogin: () => {},
   logout: () => {},
+  refreshJWT: () => {},
 }
 
 const UserContext = createContext(initialUserContext)
 
 export function UserProvider({ children }: UserProviderProps) {
+  const { t } = useLanguageProvider()
   const [state, dispatch] = useReducer(reducer, initialState)
   const userState = state
   const setUserState = dispatch
   const client = useApolloClient()
+  const { push } = useRouter()
+  const { preferences } = usePrefs()
+  const { showToast, clearAllToasts } = useToast()
+
+  let refreshTokenTimer = useRef(0)
+
+  const loginTimer = useMemo(
+    () =>
+      new LoginInactivityTimer({
+        idleTimeout: preferences.logoutAfterInactivity,
+        onLogout: () => {
+          logout()
+          showToast({
+            title: t('MENU_LOGOUT'),
+            text: t('LOGOUT_INACTIVITY_ALERT'),
+            style: 'negative',
+            position: Position.bottomLeft,
+            timeout: 0,
+          })
+        },
+      }),
+    []
+  )
 
   const logout = () => {
-    // Delete everything EXCEPT language preference in localStorage
-    const language = localStorage.getItem('language')
-    localStorage.clear()
-    if (language) localStorage.setItem('language', language)
+    clearInterval(refreshTokenTimer.current)
+    refreshTokenTimer.current = 0
+    clearLocalStorageExcept(['language', LOCAL_STORAGE_EXPIRY_KEY])
     client.clearStore()
-    window.location.href = '/login'
+    setUserState({ type: 'resetCurrentUser' })
+    loginTimer.end()
+    push('/login')
+  }
+
+  const delayedLogout = () => {
```

Comment:
This is what happens when the front-end tries to run "fetchUserInfo" with an invalid/expired token. We need to implement a more generic overall solution for what happens on various server responses/errors, but that's a bigger job for another day.